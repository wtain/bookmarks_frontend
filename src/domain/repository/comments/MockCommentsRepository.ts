import CommentDto from '../../dto/CommentDto';
import ICommentsRepository from "./ICommentsRepository";

class MockCommentsRepository implements ICommentsRepository {

  private comments: CommentDto[];

  constructor() {
    this.comments = [
      {
        id: "1",
        created: new Date(),
        bookmarkId: "1",
        contents: "Test comment 1"
      },
      {
        id: "2",
        created: new Date(),
        bookmarkId: "1",
        contents: "Test comment 2"
      },
      {
        id: "3",
        created: new Date(),
        bookmarkId: "2",
        contents: "Test comment 3"
      }
    ];
  }

  async getComments(bookmarkId: string): Promise<CommentDto[]> {
    return this.comments.filter(comment => comment.bookmarkId == bookmarkId);
  }

  async addComment(comment: CommentDto): Promise<CommentDto> {
    this.comments = [...this.comments, comment];
    return comment;
  }

  async deleteComment(id: string): Promise<void> {
    this.comments = this.comments.filter(comment => comment.id !== id);
  }

  async updateComment(comment: CommentDto): Promise<CommentDto> {
    this.comments = this.comments.map(c => c.id === comment.id ? comment : c);
    return comment;
  }

}

export default MockCommentsRepository;