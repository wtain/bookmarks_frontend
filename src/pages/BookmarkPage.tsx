import IBookmarksRepository from "../domain/repository/bookmarks/IBookmarksRepository";
import BookmarkDto from '../domain/dto/BookmarkDto';
import { useEffect, useState } from "react";
import Bookmark from "../components/UI/Bookmark/Bookmark";
import { useParams } from "react-router-dom";
import CommentList from "../components/UI/CommentList/CommentList";
import CommentDto from "../domain/dto/CommentDto";
import ICommentsRepository from '../domain/repository/comments/ICommentsRepository';
import CommentEditor from "../components/UI/CommentEditor/CommentEditor";
import {v4 as uuidv4} from 'uuid';

interface Props {
  bookmarksRepository: IBookmarksRepository;
  commentsRepository: ICommentsRepository;
}

const BookmarkPage = (props: Props) => {

  const { bookmarkId } = useParams();

  const [bookmark, setBookmark] = useState<BookmarkDto>();
  const [comments, setComments] = useState<CommentDto[]>([]);

  useEffect(() => {
    (async function () {
      const bookmark = await props.bookmarksRepository.getBookmark(bookmarkId!);
      setBookmark(bookmark);
    })()
  }, []);

  // todo: factory for comments repository? to be created from bookmarks repository??
  useEffect(() => {
    (async function () {
      const comments = await props.commentsRepository.getComments(bookmarkId!);
      setComments(comments);
    })()
  }, []);

  const doUpdatePoll = async () => {
    setComments(
      await props.commentsRepository.getComments(bookmarkId!)
    );
  };

  useEffect(() => {
    const interval = setInterval(() => doUpdatePoll(), 20 * 1000);
    return () => clearInterval(interval);
  }, [bookmarkId]);
  
  // todo: separate component for comment editor textarea
  // todo: edit comments

  if (bookmark !== undefined) {
    return (
      <div>
        <Bookmark bookmark={bookmark} showExpanded={true} />
        <CommentList bookmark={bookmark} comments={comments} />
        <CommentEditor onDone={(commentText: string) => {
          // todo: Check result
          // todo: reuse this component in bookmarkform

          (async function () {
            await props.commentsRepository.addComment({
              bookmarkId: bookmarkId!,
              contents: commentText,
              id: uuidv4(), 
              created: new Date()
            });
            doUpdatePoll();
          })();
          
        }} />
      </div>
    )
  }
  return (
    <div>(Not found)</div>
  )
  // todo: + comments list + add comment form
  // todo: snapshot test
}

export default BookmarkPage;