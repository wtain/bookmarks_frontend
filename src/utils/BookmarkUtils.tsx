
import React from "react";
import BookmarkDto from "../domain/dto/BookmarkDto";
import {v4 as uuidv4} from 'uuid';
import TagDto from "../domain/dto/TagDto";

const createBookmark = (summary: string, contents: string, tags: TagDto[]): BookmarkDto => {
    return {
                summary, 
                contents, 
                id: uuidv4(), 
                created: new Date().toLocaleString(),
                tags,
                isDone: false
            }
}

const createNewTag = (): TagDto => {
    return { name: "", id: uuidv4() }
}

const BookmarkUtils = { createBookmark, createNewTag };

export default BookmarkUtils;