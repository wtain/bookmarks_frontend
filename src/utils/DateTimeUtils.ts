import moment from "moment";

export function parseDateTimeFromMongo(datetime: Date | undefined) {
  if (datetime === undefined) {
    return undefined;
  }
  return moment(datetime.toISOString()).fromNow();
}

export function addOneMonth(date: Date): Date {
  let result = new Date(date);
  result.setMonth(result.getMonth() + 1)
  return result;
}

export function firstDayOfMonth(date: Date): Date {
  let result = new Date(date);
  result.setDate(1)
  return result;
}

export function getDatePortion(date: Date): string {
  const dateCopy = new Date(date);
  dateCopy.setUTCHours(0);
  dateCopy.setMinutes(0);
  const momentForDate = moment(dateCopy);
  return momentForDate.format('YYYY-MM-DD');
}

export function datesEqual(date1: Date, date2: Date): boolean {
  return getDatePortion(date1) === getDatePortion(date2);
}