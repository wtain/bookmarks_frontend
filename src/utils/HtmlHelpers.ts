import sanitizeHtml from "sanitize-html";

export function sanitizeText(text: string){
  return sanitizeHtml(text, {
    allowedTags: ['br', 'b', 'a', 'img']
  });
}