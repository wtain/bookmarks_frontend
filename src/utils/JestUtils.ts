
export function jestHacks() {
    // https://stackoverflow.com/questions/39830580/jest-test-fails-typeerror-window-matchmedia-is-not-a-function
    // https://github.com/facebook/create-react-app/issues/10126
    window.matchMedia = (query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })
}