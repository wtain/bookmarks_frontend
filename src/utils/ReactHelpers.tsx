
import { MutableRefObject } from "react";

const scrollToRef = (ref: MutableRefObject<null | HTMLDivElement>) => {
    // On mobile Chrome smooth is not supported, so let's 
        // make sure we're scrolled anyway
        // ref.current!.scrollIntoView(true);
    ref.current?.scrollIntoView({
        behavior: "smooth",
    });
};

const ReactHelpers = {
    scrollToRef
}

export default ReactHelpers;