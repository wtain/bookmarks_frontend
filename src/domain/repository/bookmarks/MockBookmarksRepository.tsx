
import React from "react";
import BookmarkDto from "../../dto/BookmarkDto";
import IBookmarksRepository from "./IBookmarksRepository";
import ReactHelpers from "../../../utils/ReactHelpers"
import BookmarkUtils from "../../../utils/BookmarkUtils";

class MockBookmarksRepository implements IBookmarksRepository {

    private bookmarks: BookmarkDto[];

    constructor() {
        this.bookmarks = Array.from(Array(20).keys()).map(i => 
            BookmarkUtils.createBookmark("Summary " + i, 
                "Contents " + i + "\nLorem ipsum\nLorem ipsum", 
                Array.from(Array(i).keys()).map(j => { 
                    return { ...BookmarkUtils.createNewTag(), name: "tag " + j }
                })
            ));
    }

    async getBookmark(id: string): Promise<BookmarkDto> {
        // todo: implement
        await ReactHelpers.delay(300);
        return this.bookmarks[0];
    }

    public async getBookmarks(): Promise<BookmarkDto[]> {
        await ReactHelpers.delay(300);
        return this.bookmarks;
    }

    async getBookmarksByTag(tag: string): Promise<BookmarkDto[]> {
        await ReactHelpers.delay(300);
        return this.bookmarks; // todo: implement
    }

    public async addBookmark(bookmark: BookmarkDto) {
        this.bookmarks = [...this.bookmarks, bookmark]
    }

    public async removeBookmark(bookmark: BookmarkDto) {
        this.bookmarks = this.bookmarks.filter(b => b.id != bookmark.id);
    }

    public async editBookmark(bookmark: BookmarkDto) {
        this.bookmarks = this.bookmarks.map(b => b.id === bookmark.id ? bookmark : b);
    }
}

export default MockBookmarksRepository;