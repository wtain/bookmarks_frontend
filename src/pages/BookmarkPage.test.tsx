import renderer from 'react-test-renderer';
import MockBookmarksRepository from '../domain/repository/bookmarks/MockBookmarksRepository';
import { BrowserRouter } from 'react-router-dom';
import BookmarkPage from './BookmarkPage';
import MockCommentsRepository from '../domain/repository/comments/MockCommentsRepository';
import { AppProvider } from "@shopify/polaris";
import translations from "@shopify/polaris/locales/en.json";
import { jestHacks } from '../utils/JestUtils';

jestHacks();

it('BookmarkPage renders correctly', () => {
  const bookmarksRepository = new MockBookmarksRepository();
  const commentsRepository = new MockCommentsRepository();

  const tree = renderer
    .create(
      <BrowserRouter>
        <AppProvider i18n={translations}>
          <BookmarkPage bookmarksRepository={bookmarksRepository} commentsRepository={commentsRepository} />
        </AppProvider>
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});