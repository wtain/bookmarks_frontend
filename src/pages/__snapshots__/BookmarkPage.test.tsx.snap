// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkPage renders correctly 1`] = `
Array [
  <div>
    (Not found)
  </div>,
  <div
    id="PolarisPortalsContainer"
  />,
]
`;
